packages/zoe/src/typeGuards.js

```diff
@@ -79,3 +121,231 @@ export const InvitationElementShape = M.splitRecord({
+const BundleCapShape = M.any();
+
+export const ZoeStorageManagerIKit = harden({
+  zoeServiceDataAccess: M.interface('ZoeService dataAccess', {
+    getTerms: M.call(InstanceHandleShape).returns(
+      M.split(TermsShape, M.record()),
+    ),
+    getIssuers: M.call(InstanceHandleShape).returns(IssuerKeywordRecordShape),
+    getBrands: M.call(InstanceHandleShape).returns(BrandKeywordRecordShape),
+    getInstallationForInstance: M.call(InstanceHandleShape).returns(
+      M.or(M.promise(), M.remotable('Installation')),
+    ),
+    getInvitationIssuer: M.call().returns(IssuerShape),
+
+    getBundleIDFromInstallation: M.call(M.any()).returns(M.promise()),
+    installBundle: M.call(M.any()).returns(M.promise()),
+    installBundleID: M.call(M.string()).returns(M.promise()),
+
+    getPublicFacet: M.call(InstanceHandleShape).returns(
+      M.or(M.promise(), M.remotable('PublicFacet')),
+    ),
+    getOfferFilter: M.call(InstanceHandleShape).returns(M.arrayOf(M.string())),
+    setOfferFilter: M.call(
+      InstanceHandleShape,
+      M.arrayOf(M.string()),
+    ).returns(),
+    getProposalShapeForInvitation: M.call(InvitationHandleShape).returns(
+      M.or(M.pattern(), M.undefined()),
+    ),
+  }),
+  makeOfferAccess: M.interface('ZoeStorage makeOffer access', {
+    getAssetKindByBrand: M.call(BrandShape).returns(AssetKindShape),
+    installBundle: M.call(InstanceHandleShape).returns(),
+    getInstanceAdmin: M.call(InstanceHandleShape).returns(
+      M.remotable('instanceAdmin'),
+    ),
+    getProposalShapeForInvitation: M.call(InvitationHandleShape).returns(
+      M.or(M.pattern(), M.undefined()),
+    ),
+    getInvitationIssuer: M.call().returns(IssuerShape),
+    depositPayments: M.call(ProposalShape, PaymentPKeywordRecordShape).returns(
+      M.promise(),
+    ),
+  }),
+  startInstanceAccess: M.interface('ZoeStorage startInstance access', {
+    makeZoeInstanceStorageManager: M.call(
+      M.any(),
+      InstallationShape,
+      M.any(),
+      IssuerPKeywordRecordShape,
+      InstanceHandleShape,
+      BundleCapShape,
+    ).returns(M.promise()),
+    unwrapInstallation: M.callWhen(M.eref(InstallationShape)).returns(M.any()),
+  }),
+  invitationIssuerAccess: M.interface('ZoeStorage invitationIssuer', {
+    getInvitationIssuer: M.call().returns(IssuerShape),
+  }),
+});
+
+export const ZoeServiceIKit = harden({
+  zoeService: M.interface('ZoeService', {
+    install: M.call(M.any()).returns(M.promise()),
+    installBundleID: M.call(M.string()).returns(M.promise()),
+    startInstance: M.call(M.eref(InstallationShape))
+      .optional(IssuerPKeywordRecordShape, M.any(), M.any())
+      .returns(M.promise()),
+    offer: M.call(M.eref(InvitationShape))
+      .optional(ProposalShape, PaymentPKeywordRecordShape, M.any())
+      .returns(M.promise()),
+
+    getPublicFacet: M.callWhen(M.await(InstanceHandleShape)).returns(
+      M.or(M.promise(), M.remotable('PublicFacet')),
+    ),
```

Comment:
The first simplification is that `M.ref(P)` is equivalent to `M.or(P, M.promise())` so you can always rewrite

```suggestion
    getPublicFacet: M.callWhen(M.await(InstanceHandleShape)).returns(
      M.eref(M.remotable('PublicFacet')),
    ),
```

This is purposely weaker than the TS type `ERef<P>` because the TS type also says that if the result is a promise that fulfills, it must fulfill to a P. If you would prefer this stronger TS meaning, the `M.callWhen` provides it for arguments using `M.await` as you do above. But `M.callWhen` also provides it implicitly for return values, so you could just say

```suggestion
    getPublicFacet: M.callWhen(M.await(InstanceHandleShape)).returns(
      M.remotable('PublicFacet'),
    ),
```

`M.callWhen` uniquely can provide this stronger checking because it is allowed to (and does) introduce extra turns.

Reply:
cool

Reply:
Looks done. Resolving.

Reply:
I still see some `M.callWhen(...).returns(M.or(M.promise(), ...))` that likely should be simplified to `M.callWhen(...).returns(...)`, for example on lines 313, 316, and 324.

Unresolving

Reply:
There are also some `M.or(M.promise(), ...)` not in a callWhen, and so not that simplifiable , but which likely still should be simplified to `M.eref(...)`. For example, on line 252.

Reply:
I think I got them all this time.  

[GitHub is making it hard to find all your unresolved comments.]

Reply:
> I think I got them all this time.

Yup. Resolving.

> [GitHub is making it hard to find all your unresolved comments.]

For me too!